src/main/scala/mesosphere/marathon/api/akkahttp/v2/AppsController.scala

```diff
@@ -6,90 +6,131 @@ import java.time.Clock
-          Some(authzSelector)
-        ).flatten
         val resolvedEmbed = InfoEmbedResolver.resolveApp(embed.toSet) + AppInfo.Embed.Counts + AppInfo.Embed.Deployments
-        appInfoService.selectAppsBy(Selector.forall(selectors), resolvedEmbed)
+        val selector = selectAuthorized(search(cmd, id, label))
+        appInfoService.selectAppsBy(selector, resolvedEmbed)
       }
       onSuccess(index)(apps => complete(Json.obj("apps" -> apps)))
     }
   }
 
-  private def createApp(app: AppDefinition, force: Boolean)(implicit identity: Identity): Route = {
-    def create: Future[(DeploymentPlan, AppInfo)] = {
+  private[v2] def search(cmd: Option[String], id: Option[String], label: Option[String]): AppSelector = {
+    def containCaseInsensitive(a: String, b: String): Boolean = b.toLowerCase contains a.toLowerCase
+    val selectors = Seq[Option[Selector[AppDefinition]]](
+      cmd.map(c => Selector(_.cmd.exists(containCaseInsensitive(c, _)))),
+      id.map(s => Selector(app => containCaseInsensitive(s, app.id.toString))),
+      label.map(new LabelSelectorParsers().parsed)
+    ).flatten
+    Selector.forall(selectors)
+  }
 
-      def createOrThrow(opt: Option[AppDefinition]) = opt
-        .map(_ => throw ConflictingChangeException(s"An app with id [${app.id}] already exists."))
-        .getOrElse(app)
+  private[v2] def selectAuthorized(fn: => AppSelector)(implicit identity: Identity): AppSelector = {
+    Selector.forall(Seq(authzSelector, fn))
+  }
 
-      groupManager.updateApp(app.id, createOrThrow, app.version, force).map { plan =>
-        val appWithDeployments = AppInfo(
-          app,
-          maybeCounts = Some(TaskCounts.zero),
-          maybeTasks = Some(Seq.empty),
-          maybeDeployments = Some(Seq(Identifiable(plan.id)))
-        )
-        plan -> appWithDeployments
-      }
+  private def replaceMultipleApps(implicit identity: Identity): Route = {
+    parameters('partialUpdate.?(true)) { partialUpdate =>
+      updateMultiple(partialUpdate, allowCreation = true)
     }
-    authorized(CreateRunSpec, app).apply {
-      onSuccess(create) { (plan, app) =>
-        //TODO: post ApiPostEvent
-        complete((StatusCodes.Created, Seq(Headers.`Marathon-Deployment-Id`(plan.id)), app))
+  }
+
+  private def patchMultipleApps(implicit identity: Identity): Route = {
+    updateMultiple(partialUpdate = true, allowCreation = false)
+  }
+
+  private def createApp(implicit identity: Identity): Route = {
+    (entity(as[AppDefinition]) & forceParameter & extractClientIP & extractUri) { (rawApp, force, remoteAddr, reqUri) =>
+
+      authorized(CreateRunSpec, rawApp).apply {
+        def create: Future[(DeploymentPlan, AppInfo)] = {
+
+          val app = rawApp.copy(versionInfo = VersionInfo.OnlyVersion(clock.now()))
+
+          def createOrThrow(opt: Option[AppDefinition]) = opt
+            .map(_ => throw ConflictingChangeException(s"An app with id [${app.id}] already exists."))
```

Comment:
Why not reject? So, the controllers define routes or at least return them. I have a hard time to follow the design decision here. We seem to mix Akka HTTP's routes and our own logic. Why not avoid routes here all together and e.g. pass the already unmarshalled app definition?